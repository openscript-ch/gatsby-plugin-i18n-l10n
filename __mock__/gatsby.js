const React = require('react');
const gatsby = jest.requireActual('gatsby');
module.exports = {
  ...gatsby,
  graphql: jest.fn(),
  // these props are invalid for an `a` tag
  Link: jest.fn(({ activeClassName, activeStyle, getProps, innerRef, partiallyActive, ref, replace, to, ...rest }) =>
    React.createElement('a', {
      ...rest,
      href: to,
    }),
  ),
  StaticQuery: jest.fn(),
  useStaticQuery: jest.fn(),
};
