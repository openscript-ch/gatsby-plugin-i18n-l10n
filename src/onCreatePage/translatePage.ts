import { Actions, CreatePageArgs, Page, PluginOptions } from 'gatsby';
import { SitePageContext, UnstatefulSitePageContext } from '../../types';
import { createLocalePagesId } from '../utils/i18n';
import { generatePageContextByPath, translatePagePath, translatePagePaths } from '../utils/path';

const isPagePathBlacklisted = (path: string, options: PluginOptions, locale?: string) => {
  const localeOptions = options.locales.find((l) => l.locale === locale);

  if (!localeOptions?.pageBlacklist) {
    return false;
  }

  if (localeOptions.pageBlacklist.find((pbl) => path.includes(pbl))) {
    return true;
  }

  return false;
};

const createTranslatedPage = (createPage: Actions['createPage'], page: Page<SitePageContext>, options: PluginOptions) => {
  if (!isPagePathBlacklisted(page.path, options, page.context?.locale)) {
    createPage(page);
  }
};

export const translatePage = async ({ page, actions }: CreatePageArgs<SitePageContext>, options: PluginOptions) => {
  const { createPage, deletePage } = actions;

  // If this page was already translated, we skip it.
  if (page.context?.locale && page.context.localePagesId) {
    return;
  }

  if (page.isCreatedByStatefulCreatePages) {
    // Translate statefully created pages from `/src/pages` or gatsby-plugin-page-creator
    const paths = translatePagePaths(page.path, options);

    deletePage(page);

    paths.forEach((path) => {
      const translations = paths.filter((p) => p.locale !== path.locale && !isPagePathBlacklisted(p.path, options, p.locale));
      const locale = options.locales.find((l) => l.locale === path.locale);
      const context = {
        ...page.context,
        locale: path.locale,
        localePagesId: createLocalePagesId(page.path),
        translations,
        prefix: locale?.prefix,
      };

      createTranslatedPage(createPage, { ...page, path: path.path, context }, options);
    });
  } else {
    // Translate programmically created pages
    deletePage(page);

    const { referTranslations, adjustPath, ...restContext } = (page.context as UnstatefulSitePageContext) || {};
    let context = restContext;
    let path = (context.basePath as string) || page.path;

    // Add locale, localePagesId and prefix from context or path
    const contextLocale = page.context?.locale as string | undefined;
    let optionsLocale = options.locales.find((l) => l.locale === contextLocale);
    const localeAndPrefixContext = optionsLocale
      ? { locale: optionsLocale.locale, prefix: optionsLocale.prefix }
      : generatePageContextByPath(path, options);
    context = { ...context, ...localeAndPrefixContext, localePagesId: createLocalePagesId(page.path, localeAndPrefixContext.prefix) };
    optionsLocale = options.locales.find((l) => l.locale === localeAndPrefixContext.locale);

    // Refer translations if requested
    if (referTranslations && Array.isArray(referTranslations) && referTranslations.length > 0) {
      const translations = translatePagePaths(path, options).filter(
        (p) =>
          p.locale !== localeAndPrefixContext.locale &&
          referTranslations.includes(p.locale) &&
          !isPagePathBlacklisted(p.path, options, p.locale),
      );
      context = { ...context, translations };
    }

    // Translate page path if requested
    if (adjustPath && typeof context.locale === 'string' && typeof context.prefix === 'string' && optionsLocale) {
      path = translatePagePath(path, optionsLocale.slugs, context.locale, context.prefix, options.defaultLocale);
    }

    createTranslatedPage(createPage, { ...page, context, path }, options);
  }
};
