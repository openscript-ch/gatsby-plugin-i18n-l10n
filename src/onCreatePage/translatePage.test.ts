import { Page, PluginOptions } from 'gatsby';
import { translatePage } from './translatePage';

const options: PluginOptions = {
  defaultLocale: `en-US`,
  siteUrl: '',
  locales: [
    {
      locale: `en-US`,
      prefix: `en`,
      slugs: {},
      messages: {},
    },
    {
      locale: `de-CH`,
      prefix: `de`,
      slugs: {
        '/imprint': '/impressum',
        '/pages/imprint': '/seiten/impressum',
      },
      messages: {},
      pageBlacklist: ['/do-not-translate-to-german'],
    },
    {
      locale: `zh-CN`,
      prefix: `zh`,
      slugs: {},
      messages: {},
    },
  ],
  plugins: [],
};

let actions: any;

describe('translatePage', () => {
  beforeEach(() => {
    actions = {
      createPage: jest.fn(),
      deletePage: jest.fn(),
    };
  });

  it('should translate stateful created pages', () => {
    const page: Page = {
      path: '/pages/imprint',
      component: {} as any,
      context: {},
      isCreatedByStatefulCreatePages: true,
    };
    translatePage({ page, actions } as any, options);

    expect(actions.deletePage).toHaveBeenCalledWith(page);
    expect(actions.createPage).toHaveBeenCalledTimes(3);
    expect(actions.createPage).toHaveBeenNthCalledWith(1, {
      ...page,
      context: {
        locale: 'en-US',
        prefix: 'en',
        localePagesId: 'pages.imprint',
        translations: [
          {
            locale: 'de-CH',
            path: '/de/seiten/impressum',
          },
          {
            locale: 'zh-CN',
            path: '/zh/pages/imprint',
          },
        ],
      },
    });
    expect(actions.createPage).toHaveBeenNthCalledWith(2, {
      ...page,
      context: {
        locale: 'de-CH',
        prefix: 'de',
        localePagesId: 'pages.imprint',
        translations: [
          {
            locale: 'en-US',
            path: '/pages/imprint',
          },
          {
            locale: 'zh-CN',
            path: '/zh/pages/imprint',
          },
        ],
      },
      path: '/de/seiten/impressum',
    });
    expect(actions.createPage).toHaveBeenNthCalledWith(3, {
      ...page,
      context: {
        locale: 'zh-CN',
        prefix: 'zh',
        localePagesId: 'pages.imprint',
        translations: [
          {
            locale: 'en-US',
            path: '/pages/imprint',
          },
          {
            locale: 'de-CH',
            path: '/de/seiten/impressum',
          },
        ],
      },
      path: '/zh/pages/imprint',
    });
  });

  it('should translate unstateful created pages by path with default language', () => {
    const page: Page = {
      path: '/imprint',
      component: {} as any,
      isCreatedByStatefulCreatePages: false,
    };
    translatePage({ page, actions } as any, options);

    expect(actions.deletePage).toHaveBeenCalledWith(page);
    expect(actions.createPage).toHaveBeenCalledTimes(1);
    expect(actions.createPage).toHaveBeenNthCalledWith(1, {
      ...page,
      context: {
        locale: 'en-US',
        prefix: 'en',
        localePagesId: 'imprint',
      },
    });
  });

  it('should translate unstateful created pages by prefix in path', () => {
    const page: Page = {
      path: '/de/imprint',
      component: {} as any,
      context: {},
      isCreatedByStatefulCreatePages: false,
    };
    translatePage({ page, actions } as any, options);

    expect(actions.deletePage).toHaveBeenCalledWith(page);
    expect(actions.createPage).toHaveBeenCalledTimes(1);
    expect(actions.createPage).toHaveBeenNthCalledWith(1, {
      ...page,
      context: {
        locale: 'de-CH',
        prefix: 'de',
        localePagesId: 'imprint',
      },
    });
  });

  it('should translate unstateful created pages by locale in context', () => {
    const page: Page = {
      path: '/imprint',
      component: {} as any,
      context: {
        locale: 'de-CH',
      },
      isCreatedByStatefulCreatePages: false,
    };
    translatePage({ page, actions } as any, options);

    expect(actions.deletePage).toHaveBeenCalledWith(page);
    expect(actions.createPage).toHaveBeenCalledTimes(1);
    expect(actions.createPage).toHaveBeenNthCalledWith(1, {
      ...page,
      context: {
        locale: 'de-CH',
        prefix: 'de',
        localePagesId: 'imprint',
      },
    });
  });

  it('should translate unstateful created pages and refer translations', () => {
    const page: Page = {
      path: '/imprint',
      component: {} as any,
      context: {
        locale: 'en-US',
        referTranslations: ['de-CH'],
      },
      isCreatedByStatefulCreatePages: false,
    };
    translatePage({ page, actions } as any, options);

    expect(actions.deletePage).toHaveBeenCalledWith(page);
    expect(actions.createPage).toHaveBeenCalledTimes(1);
    expect(actions.createPage).toHaveBeenNthCalledWith(1, {
      ...page,
      context: {
        locale: 'en-US',
        prefix: 'en',
        localePagesId: 'imprint',
        translations: [
          {
            locale: 'de-CH',
            path: '/de/impressum',
          },
        ],
      },
    });
  });

  it('should translate unstateful created pages and adjust the path', () => {
    const page: Page = {
      path: '/imprint',
      component: {} as any,
      context: {
        locale: 'de-CH',
        referTranslations: ['en-US', 'zh-CN'],
        adjustPath: true,
      },
      isCreatedByStatefulCreatePages: false,
    };
    translatePage({ page, actions } as any, options);

    expect(actions.deletePage).toHaveBeenCalledWith(page);
    expect(actions.createPage).toHaveBeenCalledTimes(1);
    expect(actions.createPage).toHaveBeenNthCalledWith(1, {
      ...page,
      path: '/de/impressum',
      context: {
        locale: 'de-CH',
        prefix: 'de',
        localePagesId: 'imprint',
        translations: [
          {
            locale: 'en-US',
            path: '/imprint',
          },
          {
            locale: 'zh-CN',
            path: '/zh/imprint',
          },
        ],
      },
    });
  });

  it('should skip already translated pages', () => {
    const page: Page = {
      path: '/imprint',
      component: {} as any,
      context: {
        locale: 'de-CH',
        localePagesId: 'imprint',
      },
    };

    translatePage({ page, actions } as any, options);

    expect(actions.deletePage).not.toHaveBeenCalled();
    expect(actions.createPage).not.toHaveBeenCalled();
  });

  it('should obey page blacklists', () => {
    const page: Page = {
      path: '/do-not-translate-to-german',
      component: {} as any,
      context: {},
      isCreatedByStatefulCreatePages: true,
    };

    translatePage({ page, actions } as any, options);

    expect(actions.createPage).toMatchSnapshot();
  });
});
