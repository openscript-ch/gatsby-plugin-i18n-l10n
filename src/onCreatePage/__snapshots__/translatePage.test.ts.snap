// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translatePage should obey page blacklists 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "component": {},
        "context": {
          "locale": "en-US",
          "localePagesId": "do-not-translate-to-german",
          "prefix": "en",
          "translations": [
            {
              "locale": "zh-CN",
              "path": "/zh/do-not-translate-to-german",
            },
          ],
        },
        "isCreatedByStatefulCreatePages": true,
        "path": "/do-not-translate-to-german",
      },
    ],
    [
      {
        "component": {},
        "context": {
          "locale": "zh-CN",
          "localePagesId": "do-not-translate-to-german",
          "prefix": "zh",
          "translations": [
            {
              "locale": "en-US",
              "path": "/do-not-translate-to-german",
            },
          ],
        },
        "isCreatedByStatefulCreatePages": true,
        "path": "/zh/do-not-translate-to-german",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
